### DISCUSSION TITLE
What happens if you call an async wasm function while the wasm vm is busy?

### DISCUSSION THREAD

**Discussion by Amatewasu** [emscripten-core/emscripten#18747] [Q&A] [answered]
Hi, first of all thanks for emscripten, it is so great!

What happens if we call an asynchronous wasm function while the wasm VM is busy?

For example, consider you have two C functions named `heavy_stuff_1` and `heavy_stuff_2` and we compile it to wasm with emscripten with the asyncify option:
`emcc myFile.c -sASYNCIFY -sASYNCIFY_IMPORTS=heavy_stuff_1,heavy_stuff_2`.
with the following JavaScript code:
```js
MyWasmModule.ccall('heavy_stuff_1', [], [], [], { async: true });
MyWasmModule.ccall('heavy_stuff_2', [], [], [], { async: true });
```

I mean, the `heavy_stuff_1` will not have returned yet when we will call `heavy_stuff_2`. Is it ok do to so?

If yes we could do this:
```js
await Promise.all([MyWasmModule.ccall('heavy_stuff_1', [], [], [], { async: true }), MyWasmModule.ccall('heavy_stuff_2', [], [], [], { async: true })]);
```

Thank you for your time and have a great day!

**Comment by sbc100** [comment, marked as answer]
IIUC this doesn't work, and asyncify only supports one suspended task at a time.  Is that correct @kripken?   Do we detect this type of thing in debug builds and report an error?

**Reply by Amatewasu** [reply]
Thank you for your _super fast_ answer. So if I understand correctly, the second call will crash instantly. So we need to take care that we never call a function while the VM is busy, right?

**Reply by kripken** [reply]
Yes, in a debug build this should error. We don't support multiple async operations at the same time (except when using the Fibers API, which is a safe way to do it).

Yes, you'd need to make sure to avoid such situations. Note that a non-async operation is ok though (like you can call into the wasm to do a malloc even when it is in the middle of an async operation).

**Reply by Amatewasu** [reply]
Alright, super interesting. Thank you for your answers. I hope it will help other people.
